python/pyarrow/table.pxi

```diff
@@ -845,7 +865,14 @@ cdef class Table:
         if nthreads is None:
             nthreads = cpu_count()
 
-        mgr = pdcompat.table_to_blockmanager(self, nthreads)
+        table = self
+        for ix, col in enumerate(self.itercolumns()):
+            if str_to_categorical and col.type.id == Type_STRING:
+                # Doesn't matter what type the dictionary is initialized with
+                dtype = dictionary(col.type, col.data.chunk(0))
+                table = table.cast_column(ix, dtype)
+
+        mgr = pdcompat.table_to_blockmanager(table, nthreads)
```

Comment:
I think all this logic needs to push down into `arrow/python/arrow_to_pandas.cc`:

* Will not require as much temporary memory allocation (here all of the columns are casted prior to conversion)
* Conversions can be performed in parallel